packages/amplify-graphql-api-construct/src/internal/authorization-modes.ts

```diff
@@ -0,0 +1,134 @@
+import { AppSyncAuthConfiguration, AppSyncAuthConfigurationEntry } from '@aws-amplify/graphql-transformer-interfaces';
+import {
+  ApiKeyAuthorizationConfig,
+  AuthorizationConfig,
+  IAMAuthorizationConfig,
+  LambdaAuthorizationConfig,
+  OIDCAuthorizationConfig,
+  UserPoolAuthorizationConfig,
+} from '../types';
+
+type AuthorizationConfigMode =
+  | IAMAuthorizationConfig & { type: 'AWS_IAM' }
+  | UserPoolAuthorizationConfig & { type: 'AMAZON_COGNITO_USER_POOLS' }
+  | OIDCAuthorizationConfig & { type: 'OPENID_CONNECT' }
+  | ApiKeyAuthorizationConfig & { type: 'API_KEY' }
+  | LambdaAuthorizationConfig & { type: 'AWS_LAMBDA' };
+
+/**
+ * Converts a single auth mode config into the amplify-internal representation.
+ * @param authMode the auth mode to convert into the Appsync CDK representation.
+ */
+const convertAuthModeToAuthProvider = (authMode: AuthorizationConfigMode): AppSyncAuthConfigurationEntry => {
+  const authenticationType = authMode.type;
+  switch (authMode.type) {
+    case 'API_KEY': return {
+      authenticationType,
+      apiKeyConfig: {
+        description: authMode.description,
+        apiKeyExpirationDays: authMode.expires.toDays(),
+      },
+    };
+    case 'AWS_IAM': return { authenticationType };
+    case 'AMAZON_COGNITO_USER_POOLS': return {
+      authenticationType,
+      userPoolConfig: {
+        userPoolId: authMode.userPool.userPoolId,
+      },
+    };
+    case 'OPENID_CONNECT': return {
+      authenticationType,
+      openIDConnectConfig: {
+        name: authMode.oidcProviderName,
+        issuerUrl: authMode.oidcIssuerUrl,
+        clientId: authMode.clientId,
+        iatTTL: authMode.tokenExpiryFromIssue.toSeconds(),
+        authTTL: authMode.tokenExpiryFromAuth.toSeconds(),
+      },
+    };
+    case 'AWS_LAMBDA': return {
+      authenticationType,
+      lambdaAuthorizerConfig: {
+        lambdaFunction: authMode.function.functionName,
+        ttlSeconds: authMode.ttl.toSeconds(),
+      },
+    };
+    default: throw new Error(`Unexpected AuthMode type ${authenticationType} encountered.`);
+  }
+};
+
+/**
+ * Given an appsync auth configuration, convert into appsync auth provider setup.
+ * @param authConfig the config to transform
+ * @returns the appsync config object.
+ */
+const convertAuthConfigToAppSyncAuth = (authConfig: AuthorizationConfig): AppSyncAuthConfiguration => {
+  // Convert auth modes into an array of appsync configs, and include the type so we can use that for switching and partitioning later.
+  const authModes = [
+    authConfig.apiKeyConfig ? { type: 'API_KEY', ...authConfig.apiKeyConfig } : null,
```

Comment:
Currently, it's not feasible with the constraints of the underlying transformer system, which is why I've refactored this to be a map instead of list of modes.